### PR TITLE
Rest Client: Clean up search objects

### PR DESCRIPTION
**Pull request by javanna**
This PR includes a bunch of changes discussed after the initial commit of the high level client feature branch.

### CODE REVIEW

File: client/highlevel-rest/src/main/java/org/elasticsearch/client/SearchHit.java [hunk 1]
```diff
@@ -21,185 +21,97 @@
...
 import java.util.List;
 import java.util.Map;
 import java.util.Map.Entry;
 
-public class ClientSearchHit implements SearchHit {
+public class SearchHit implements Iterable<SearchHitField> {
 
-    private Map<String, Object> hit;
-    private XContentAccessor objectPath;
+    private final XContentAccessor objectPath;
```
Comment (tlrx) [inline client/highlevel-rest/src/main/java/org/elasticsearch/client/SearchHit.java]: If we don't move this outside the class let's at least rename it to `accessor`?

File: client/highlevel-rest/src/main/java/org/elasticsearch/client/SearchHit.java [hunk 2]
```diff
@@ -21,185 +21,97 @@
...
-    }
-
-    @Override
-    public byte[] source() {
-        return BytesReference.toBytes(sourceRef());
-    }
-
-    @Override
+    //TODO add hasElement support to XContentAccessor?
     public boolean hasSource() {
```
Comment (tlrx) [inline client/highlevel-rest/src/main/java/org/elasticsearch/client/SearchHit.java]: I'd rather only propose `getSource()` that returns null when no response body, an empty map when the response body is empty and a non-empty map that reflects the response body.

File: client/highlevel-rest/src/main/java/org/elasticsearch/client/SearchHit.java [hunk 3]
```diff
@@ -21,185 +21,97 @@
...
         return (Map<String, Object>) source;
     }
 
-    @Override
-    public String sourceAsString() {
-        return mapToString(getSource());
-    }
-
-    @Override
+    //TODO all these getSourceAs* methods are a bit misleading as after all they are all based on the map that we have already parsed into
```
Comment (tlrx) [inline client/highlevel-rest/src/main/java/org/elasticsearch/client/SearchHit.java]: I think we should just not add those methods

File: client/highlevel-rest/src/main/java/org/elasticsearch/client/SearchHits.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+public class SearchHits implements Iterable<SearchHit> {
+
+    private final XContentAccessor objectPath;
+    private SearchHit[] hitsArray;
```
Comment (tlrx) [inline client/highlevel-rest/src/main/java/org/elasticsearch/client/SearchHits.java]: renamed to `hits`?

File: client/highlevel-rest/src/main/java/org/elasticsearch/client/SearchHits.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+ */
+
+package org.elasticsearch.client;
+
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+public class SearchHits implements Iterable<SearchHit> {
```
Comment (tlrx) [inline client/highlevel-rest/src/main/java/org/elasticsearch/client/SearchHits.java]: Javadoc?

**Comment by javanna** [comment]
Closing this, we discussed and we decided to not go down this path of duplicating requests and response objects from the java api.

**Pull request closed without merging**